doc/compiler/optimizer/DataFlowEngine.md

```diff
@@ -0,0 +1,123 @@
+
+## Virtual Hierarchy
+
+The entire data-flow engine is based on virtual inheritance, in which each class builds upon its parent to provide additional functionality and the leaf class is typically the concrete data-flow analysis one may find in literature. Tracing a path through the virtual hierarchy will build different analyses based off of what information (sets) we wish to be flown around the program, in which manner (forward & backward), and how we wish information to be aggregated (union & intersection) as the data-flow engine traverses the CFG.
+
+To aid in understanding how this virtual hierarchy works let's consider the following simple program:
+
+```
+int y = 5;
+int z = 0;
+
+for (int x = 0; x < 10; ++x) {
+    z = z + y;
+}
+
+print(z);
+```
+
+It is trivial for a human to deduce that the variable `y` is not modified within the loop, and that there only exists one possible definition of `y` which reaches the use inside the loop. Therefore we should be able to propagate the right hand side of the definition of `y` to the use inside the loop, meaning we can simplify the program as:
+
+```
+int z = 0;
+
+for (int x = 0; x < 10; ++x) {
+    z = z + 5;
+}
+
+print(z);
+```
+
+For a compiler to be able to deduce this on arbitrary programs it must use a data-flow analysis called _reaching definitions_ which for a particular _use_ of a variable at a particular point in the program gives the list of all possible definitions (_def_) of that variable that can reach the particular use.
+
+[ReachingDefinitions.cpp](https://github.com/eclipse/omr/blob/f2bc4f8f6eb09f6cc8fc4ba48717de4880b970e3/compiler/optimizer/ReachingDefinitions.cpp) implements the reaching definitions data-flow analysis. We will use this as an example to explain the virtual hierarchy of classes which drive data-flow analyses.
+
+### [`TR_DataFlowAnalysis`](https://github.com/eclipse/omr/blob/f2bc4f8f6eb09f6cc8fc4ba48717de4880b970e3/compiler/optimizer/DataFlowAnalysis.hpp#L98-L223)
+
+The root of the hierarchy that all data-flow analyses start at is `TR_DataFlowAnalysis`. This class contains mostly boilerplate code that aids in memory allocation, access to commonly used data structures, aliasing questions, tracing, and debugging information. This class also provides pure virtual functions which must be implemented by the derived classes. Developers writing new data-flow analyses will typically not be inheriting this class.
+
+### [`TR_BasicDFSetAnalysis<T>`](https://github.com/eclipse/omr/blob/f2bc4f8f6eb09f6cc8fc4ba48717de4880b970e3/compiler/optimizer/DataFlowAnalysis.hpp#L225-L359)
+
+The `TR_BasicDFSetAnalysis<T>` inherits from `TR_DataFlowAnalysis` and is a templated class representing data-flow analyses which use containers (sets) of bits which are flown around the program during the analysis. The set is represented by the template parameter `T` and it can be either a `TR_BitVector` or bit set (for example `TR_SingleBitContainer`).
```

Comment:
Single bit containers are only mentioned here with no further explanation. Wanted to keep it high level enough and not going into too many specifics of particular data-flow analyses. It may be best if OSR analysis is elaborated in the OSR document? Not sure what the best way to go about this is.

Reply:
mm I see; happy with whatever you think best since I'm close to the code so I'm fine if you think none of those details are worth including here. I believe there is some discussion of them already in the OSR documentation. The details of the single bit and tradeoff of structure vs fake structure is what I was thinking might be of interest. Happy to defer to your judgement.

Reply:
I think we'll leave things as is and leave this review in the PR as unresolved for future reference.